Add example with Ethers-EXT for Klaytn

nohkwak opened the pull request:
- FeeDelegation example added 
- Add dependency @Klaytn/ethers-ext

nohkwak committed:
Add FeePayer example

nohkwak committed:
Add ethers-ext dependency

ian0371 reviewed (changes requested):
Please also add "package-lock.json"

kjeom commented:
@nohkwak When I tested, the gasLimit is set by float number.
Please check it again. I think it needs to be casted to int in ethers-ext

```shell
Using contract at: 0xa9eF4a5BfB21e92C06da23Ed79294DaB11F5A6df
number before increment: BigNumber { value: "201" }
{
  type: 49,
  to: '0xa9eF4a5BfB21e92C06da23Ed79294DaB11F5A6df',
  value: 0,
  gasPrice: '0xba43b7400',
  gasLimit: 40084.5,
  from: '0xA2a8854b1802D8Cd5De631E690817c253d6a9153',
  nonce: 213,
  chainId: 1001,
  input: '0xd09de08a'
}
Error: underflow [ See: https://links.ethers.org/v5-errors-NUMERIC_FAULT-underflow ] (fault="underflow", operation="BigNumber.from", value=40084.5, code=NUMERIC_FAULT, version=bignumber/5.7.0)
    at Logger.makeError (/Users/contract-template/node_modules/@ethersproject/logger/src.ts/index.ts:269:28)
    at Logger.throwError (/Users/contract-template/node_modules/@ethersproject/logger/src.ts/index.ts:281:20)
    at throwFault (/Users/contract-template/node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:356:19)
    at Function.BigNumber.from (/Users/contract-template/node_modules/@ethersproject/bignumber/src.ts/bignumber.ts:243:17)
    at FieldTypeNumberBits.canonicalize (/Users/contract-template/node_modules/@klaytn/ethers-ext/src/core/field.ts:190:26)
    at /Users/contract-template/node_modules/@klaytn/ethers-ext/src/core/field.ts:273:39
    at /Users/contract-template/node_modules/lodash/lodash.js:4967:15
    at baseForOwn (/Users/contract-template/node_modules/lodash/lodash.js:3032:24)
    at Function.forOwn (/Users/contract-template/node_modules/lodash/lodash.js:13082:24)
    at TxTypeFeeDelegatedSmartContractExecution.setFields (/Users/contract-template/node_modules/@klaytn/ethers-ext/src/core/field.ts:269:7) {
  reason: 'underflow',
  code: 'NUMERIC_FAULT',
  fault: 'underflow',
  operation: 'BigNumber.from',
  value: 40084.5
}
```

nohkwak commented:
@kjeom 
The underflow error that you advised above is reflected in https://github.com/klaytn/web3klaytn/pull/418. 
Thanks.

nohkwak committed:
Update counter_inc_feePayer.ts

nohkwak committed:
Update package-lock.json

nohkwak committed:
Update package

ian0371 commented on script/counter_inc_feePayer.ts:
```diff
@@ -0,0 +1,97 @@
+import { Wallet as KlaytnWallet } from "@klaytn/ethers-ext";
+import { normalizeHardhatNetworkAccountsConfig } from "hardhat/internal/core/providers/util";
+
+async function getKlaytnSigners(): KlaytnWallet[] {
```
How about moving this function to [hardhat-utils](https://github.com/klaytn/hardhat-utils) in the future PR?? If so, should we create a `util` directory there?  @kjeom @blukat29 

We'll just merge this PR for now, and remove this function in the future PR after hardhat-utils supports it.

blukat29 replied:
Oh, hardhat-utils is the place! I've been puzzled to find where to put this.
I strongly agree with your idea.

ian0371 replied:
hardhat-utils v0.0.7 has been released. Please update like https://github.com/klaytn/hardhat-utils/pull/10#issuecomment-1628583304 @nohkwak

nohkwak replied:
Yes, I updated. (https://github.com/klaytn/contract-template/pull/1/commits/cc15f51f936dfd085bf07c13bb0fff3da9997fa9)

ian0371 commented on script/counter_inc_feePayer.ts:
```diff
@@ -0,0 +1,97 @@
+  decodedTx.feePayer = await feePayer.getAddress();
+  console.log(decodedTx);
+
+  const sentTx = await feePayer.sendTransactionAsFeePayer(decodedTx);
```
Let's remove three console logs:
```
  console.log(decodedTx);
  console.log("sentTx", sentTx);
  console.log("receipt", rc);
```

and add one:
```
  console.log("sending FeePayer transaction", "sender", sender.address, "feePayer", feePayer.address);
```

nohkwak replied:
Yes, I updated

nohkwak committed:
Update counter_inc_feePayer.ts

nohkwak committed:
Update example according to version upgrade of hardhat-utils

nohkwak committed:
Update hardhat.config.ts

nohkwak committed:
Update counter_inc_feePayer.ts

blukat29 commented:
After https://github.com/klaytn/hardhat-utils/pull/12 is merged and published,
bump hardhat-utils to 0.0.8, and then retry, please.

ian0371 commented on hardhat.config.ts:
```diff
@@ -3,10 +3,10 @@ import "hardhat-deploy";
 // the first key of test-junk
-const defaultKey = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
+const defaultKey = "0x0e4ca6d38096ad99324de0dde108587e5d7c600165ae4cd6c2462c597458c2b8";
+const defaultKey2 = "0x9435261ed483b6efa3886d6ad9f64c12078a0e28d8d80715c773e16fc000cff4";
```
please use private keys from `npx hardhat node`

nohkwak replied:
Thanks, updated.

ian0371 commented on package.json:
```diff
@@ -39,7 +39,8 @@
   "devDependencies": {
-    "@klaytn/hardhat-utils": "^0.0.5",
+    "@klaytn/ethers-ext": "^0.9.0-beta",
+    "@klaytn/hardhat-utils": "^0.0.7",
```
hardhat-utils should be 0.0.8

nohkwak replied:
Thanks, updated.

nohkwak committed:
Update config and klaytn-utils version

nohkwak committed:
Update counter_inc_feePayer.ts

@Klaytn/ethers-ext the latest version hasn't been updated yet, so I'm putting it back for a while.

blukat29 commented on script/counter_inc_feePayer.ts:
```diff
@@ -0,0 +1,56 @@
+  // }
+
+  const tx = {
+    type: 0x31,
```
TxType.FeeDelegatedSmartContractExecution

nohkwak replied:
This can be reflected after ethers-ext 0.9.2-beta is released.

blukat29 commented on script/counter_inc_feePayer.ts:
```diff
@@ -0,0 +1,56 @@
+
+  const sentTx = await feePayer.sendTransactionAsFeePayer(decodedTx);
+  await sentTx.wait();
+  console.log("sending FeePayer transaction", "sender", sender.address, "feePayer", feePayer.address);
```
Fee delegated transaction

nohkwak replied:
Yes, I updated.

blukat29 reviewed (approved):
LGTM except minor comments

nohkwak committed:
Update counter_inc_feePayer.ts
